### DISCUSSION TITLE
Translators API issues

### DISCUSSION THREAD

**Discussion by andwan0** [SubtitleEdit/subtitleedit#7702] [Q&A]
Is there any SubtitleEdit written tutorial/help/pages regarding setting up Translator APIs for the layperson?

I did google on how to create Google Translate V2 & ChatGPT API keys.

However I am getting errors.

![2023-12-03 17_56_15-](https://github.com/SubtitleEdit/subtitleedit/assets/3704766/0e83eac3-e00f-491d-b02f-a03d8881b7e4)
![2023-12-03 17_56_36-Auto-translate](https://github.com/SubtitleEdit/subtitleedit/assets/3704766/c165106b-d403-45be-a12d-e7b03c68968c)

ChatGPT error:
```
Response status code does not indicate success: 429 (Too Many Requests).
   at System.Net.Http.HttpResponseMessage.EnsureSuccessStatusCode()
   at Nikse.SubtitleEdit.Core.AutoTranslate.ChatGptTranslate.<Translate>d__12.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Nikse.SubtitleEdit.Forms.Translate.MergeAndSplitHelper.<MergeAndTranslateIfPossible>d__0.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Nikse.SubtitleEdit.Forms.Translate.AutoTranslate.<buttonTranslate_Click>d__21.MoveNext()
```

Google error:
```
Perhaps billing is not enabled (or API key is invalid)?
at Nikse.SubtitleEdit.Core.AutoTranslate.GoogleTranslateV2.Translate(String text, String sourceLanguageCode, String targetLanguageCode)
at Nikse.SubtitleEdit.Forms.Translate.AutoTranslate.<buttonTranslate_Click>d__21.MoveNext()
```

Oh and yes for Google I had to enter my credit card details to enable the Cloud services & get £300 free credit.

**Comment by niksedk** [comment]
Do try the latest beta which might have better error handling - https://github.com/SubtitleEdit/subtitleedit/releases/download/4.0.2/SubtitleEditBeta.zip
Also check the file `error_log.txt` in the SE data folder.

ChatGPT `Too many request` is clearly about wanting your credit card,

You can also install a local translation api, like libre translate (unlimited and completely free - even works in Tool - Batch convert):
See https://www.nikse.dk/subtitleedit/help#translation

My guess is Google V2 is about credit card too... I will try to test it...

**Comment by andwan0** [comment]
I just used ChatGPT API for a game:
Hysteria in Howlsbend (
https://store.epicgames.com/en-US/p/hysteria-in-howlsbend-bfaf38 )
where we have conversations with game characters and the responses are
derived from ChatGPT API usage. This works fine no errors. I didn't use a credit card for ChatGPT.

The current ChatGPT errors from Subtitle Edit state it's using the API too
fast. Couldn't you try to find a way to limit/throttle it for the "free
version" of ChatGPT?

For Google API, am currently using the "free credit" we get when we add our
credit card details. However, billing isn't enabled (ie. when my free
credit runs it, it won't use my credit card automatically).


Running ChatGPT:
-----------------------------------------------------------------------------

> Date: 12/03/2023 20:36:06
> SE: 4.0.2.0 - Microsoft Windows NT 10.0.19045.0 - 64-bit
>
> System.Net.Http.HttpRequestException
> Message: Response status code does not indicate success: 429 (Too Many
> Requests).
> StackTrace:    at
> System.Net.Http.HttpResponseMessage.EnsureSuccessStatusCode()
>    at
> Nikse.SubtitleEdit.Core.AutoTranslate.ChatGptTranslate.<Translate>d__12.MoveNext()
> --- End of stack trace from previous location where exception was thrown
> ---
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at
> System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task
> task)
>    at
> Nikse.SubtitleEdit.Forms.Translate.MergeAndSplitHelper.<MergeAndTranslateIfPossible>d__0.MoveNext()
> --- End of stack trace from previous location where exception was thrown
> ---
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at
> Nikse.SubtitleEdit.Forms.Translate.AutoTranslate.<buttonTranslate_Click>d__21.MoveNext()
>

Running Google API:
-----------------------------------------------------------------------------

> Date: 12/03/2023 20:34:21
> SE: 4.0.2.0 - Microsoft Windows NT 10.0.19045.0 - 64-bit
>
> Nikse.SubtitleEdit.Core.Translate.Service.TranslationException
> Message: "Perhaps billing is not enabled (or API key is invalid)?"
> StackTrace:    at
> Nikse.SubtitleEdit.Core.AutoTranslate.GoogleTranslateV2.Translate(String
> text, String sourceLanguageCode, String targetLanguageCode)
>    at
> Nikse.SubtitleEdit.Forms.Translate.AutoTranslate.<buttonTranslate_Click>d__21.MoveNext()
>

END OF FILE error_log.txt

**Comment by niksedk** [comment]
Free ChatGPT allows 3 calls per minute... perhaps a retry button - but it will be very slow.

I'm pretty sure that Google V2 requires billing enabled, sorry.

**Reply by francissjr** [reply]
Perhaps adding a parameter that can configure Subtitleedit and limit the request to every 20s would help.

**Comment by darnn** [comment]
I will just say again that I only pay for GPT4 in ChatGPT itself, not for the API (so only $20 a month), and that works well enough for me. I use `export text` to get something like this:
‎
`the 71 Rowsell's Street office of casing,‎`

`‎printer and publisher of magazines.‎`

`‎Jacob casing his wife Esperanza and sisters Susanna and Marianna‎`

`‎took their own lives in I Moda Harbor on the 15th of May 1940,‎`

And use the following prompt:
`The following is text from subtitles. Translate it into text for English subtitles, don't use text from previous prompts or say "here's the translation". If a line starts mid-sentence, don't capitalize the start of it:`

(Replace English with the language you need, naturally, I just find it works best translating into English.)

Then I use `import plain text` to get it back into a subtitle format, and then I use `column paste` to paste it into a fresh translation file in translation mode. Voila. You can only do about 3,000 characters at a time, but I'd say that's a minor inconvenience.

**Reply by niksedk** [reply]
@darnn: SE has a feature called `Auto-translate via copy-paste`... will probably save you a few clicks.

**Comment by dgoryeo** [comment]
I get the `"model not found"` error when I try to use ChatGPT with my key. 
I have credit in my openai account --as well as Plus subscription. So the key should work fine. 
Also please note that my api key shows up in the error message --which surprised me.
I wonder if there is a bug in the API call? Or am I doing something wrong?
I am using `version 4.0.4`



```

Response status code does not indicate success: 404 (Not Found).
   at System.Net.Http.HttpResponseMessage.EnsureSuccessStatusCode()
   at Nikse.SubtitleEdit.Core.AutoTranslate.ChatGptTranslate.<Translate>d__18.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Nikse.SubtitleEdit.Forms.Translate.MergeAndSplitHelper.<MergeAndTranslateIfPossible>d__4.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Nikse.SubtitleEdit.Forms.Translate.AutoTranslate.<buttonTranslate_Click>d__32.MoveNext()
{
    "error": {
        "message": "The model `***my api key is here***` does not exist or you do not have access to it.",
        "type": "invalid_request_error",
        "param": null,
        "code": "model_not_found"
    }
}


```

**Reply by impubes** [reply]
i am experiencing same problem as well..
